src/multivariate/jointorderstatistics.jl

```diff
@@ -0,0 +1,156 @@
+# Implementation based on chapters 2-4 of
+# Arnold, Barry C., Narayanaswamy Balakrishnan, and Haikady Navada Nagaraja.
+# A first course in order statistics. Society for Industrial and Applied Mathematics, 2008.
+
+"""
+    JointOrderStatistics <: ContinuousMultivariateDistribution
+
+The joint distribution of a subset of order statistics from a sample from a continuous
+univariate distribution.
+
+    JointOrderStatistics(
+        dist::ContinuousUnivariateDistribution,
+        n::Int,
+        ranks=Base.OneTo(n);
+        check_args::Bool=true,
+    )
+
+Construct the joint distribution of order statistics for the specified `ranks` from an IID
+sample of size `n` from `dist`.
+
+The ``i``th order statistic of a sample is the ``i``th element of the sorted sample.
+For example, the 1st order statistic is the sample minimum, while the ``n``th order
+statistic is the sample maximum.
+
+`ranks` must be a sorted vector or tuple of unique `Int`s between 1 and `n`.
+
+For a single order statistic, use [`OrderStatistic`](@ref) instead.
+
+## Examples
+
+```julia
+JointOrderStatistics(Normal(), 10)           # Product(fill(Normal(), 10)) restricted to ordered vectors
+JointOrderStatistics(Cauchy(), 10, 2:9)      # joint distribution of all but the extrema
+JointOrderStatistics(Cauchy(), 10, (1, 10))  # joint distribution of only the extrema
+```
+"""
+struct JointOrderStatistics{
+    D<:ContinuousUnivariateDistribution,R<:Union{AbstractVector{Int},Tuple{Int,Vararg{Int}}}
+} <: ContinuousMultivariateDistribution
+    dist::D
+    n::Int
+    ranks::R
+    function JointOrderStatistics(
+        dist::ContinuousUnivariateDistribution,
+        n::Int,
+        ranks::Union{AbstractVector{Int},Tuple{Int,Vararg{Int}}}=Base.OneTo(n);
+        check_args::Bool=true,
+    )
+        @check_args(
+            JointOrderStatistics,
+            (n, n ≥ 1, "`n` must be a positive integer."),
+            (
+                ranks,
+                _are_ranks_valid(ranks, n),
+                "`ranks` must be a sorted vector or tuple of unique integers between 1 and `n`.",
+            ),
+        )
+        return new{typeof(dist),typeof(ranks)}(dist, n, ranks)
+    end
+end
+
+_islesseq(x, y) = isless(x, y) || isequal(x, y)
+
+function _are_ranks_valid(ranks, n)
+    first(ranks) ≥ 1 || return false
+    last(ranks) ≤ n || return false
+    # combined with the above checks, this is equivalent to but faster than
+    # issorted(ranks) && allunique(ranks)
+    return issorted(ranks; lt=_islesseq)
+end
+function _are_ranks_valid(ranks::AbstractRange, n)
+    first(ranks) ≥ 1 || return false
+    last(ranks) ≤ n || return false
+    return issorted(ranks) && allunique(ranks)
+end
+
+length(d::JointOrderStatistics) = length(d.ranks)
+function insupport(d::JointOrderStatistics, x::AbstractVector)
+    return length(d) == length(x) && issorted(x) && all(Base.Fix1(insupport, d.dist), x)
```

Comment:
Also here, I think one loop over `x` would be sufficient, e.g.:
```suggestion
    length(d) == length(x) || return false
    xi, state = iterate(x) # at least one element!
    dist = d.dist
    insupport(dist, xi) || return false
    while (xj_state = iterate(x, state)) !== nothing
        xj, state = xj_state
        xj ≥ xi && insupport(dist, xj) || return false
        xi = xj
    end
    return true       
```